### PR TITLE
v1.4.0

### PR DESCRIPTION
**Pull request by thiagocarvalhodev**
### Release Notes
Now it's possible to save files as Stream on Android, iOS, and Web.
On the Web, it uses `FileSystemAccess` when the browser supports it, otherwise, it will use the `StreamSaver`. 

#### Limitations
- Max download on **Safari**: 1GiB

### CODE REVIEW

File: lib/src/ardrive_downloader.dart [hunk 1]
```diff
@@ -46,24 +46,24 @@ class ArDriveDownloader {
...
+    port.listen((dynamic data) {
+      DownloadTaskStatus status = DownloadTaskStatus.values[data[1]];
       int progress = data[2];
 
       /// only track the progress of current task id
       if (status == DownloadTaskStatus.enqueued) {
         controller.sink.add(0);
       } else if (status == DownloadTaskStatus.running) {
-        debugPrint('Download progress: ' + progress.toString());
+        debugPrint('Download progress: $progress');
```
Comment (karlprieb) [inline lib/src/ardrive_downloader.dart:66]: can we get rid of this log?

File: lib/src/ardrive_io.dart [hunk 2]
```diff
@@ -9,6 +9,18 @@ import 'mobile/mobile_io.dart';
 import 'web/stub_web_io.dart' // Stub implementation
     if (dart.library.html) 'web/web_io.dart';
 
+class SaveStatus {
+  final int bytesSaved;
+  final int totalBytes;
+  final bool? saveResult;
```
Comment (matibat) [inline lib/src/ardrive_io.dart:15]: ~Does the null value have a specific meaning here? Is it different than defaulting to false?~

**Edit -** Having it be null means the download hasn't yet finished.

File: lib/src/mobile/mobile_io.dart [hunk 3]
```diff
@@ -64,9 +67,19 @@ class MobileIO implements ArDriveIO {
   }
 
   @override
-  Future<void> saveFile(IOFile file) async {
+  Future<void> saveFile(IOFile file, [bool saveOnAppDirectory = false]) async {
     try {
-      await _fileSaver.save(file);
+      await _fileSaver.save(file, saveOnAppDirectory: saveOnAppDirectory);
```
Comment (matibat) [inline lib/src/mobile/mobile_io.dart:72]: We can get rid of this try..catch because it's rethrowing and nothing else.

Comment (matibat) [reply lib/src/mobile/mobile_io.dart:72]: Same comment for the method below.

File: lib/src/ardrive_io.dart (hunk 2 above)
Comment (matibat) [reply lib/src/ardrive_io.dart:15]: We can rename it to `hasFinished` or `isResultSaved`, or similar.

File: lib/src/utils/path_utils.dart [hunk 4]
```diff
@@ -111,3 +117,53 @@ Future<String> _getDefaultAndroidDir() async {
...
+
+    if (fileExtension.isNotEmpty) {
+      potentialFileName += '.$fileExtension';
+    }
+
+    testFile = File(saveDir + potentialFileName);
+
+    counter++;
+
+    // TODO: Throw an exception on arbitrarily large counter?
```
Comment (matibat) [inline lib/src/utils/path_utils.dart:159]: Bump.

File: lib/src/web/web_io.dart [hunk 5]
```diff
@@ -59,7 +63,145 @@ class WebIO implements ArDriveIO {
...
+        bytesSaved += chunk.length;
+        yield SaveStatus(
+          bytesSaved: bytesSaved,
+          totalBytes: totalBytes,
+        );
+      }
+      await writer.readyFuture;
+      debugPrint('writer ready');
+
+      // FIXME: this never ends on Firefox/Safari.
```
Comment (matibat) [inline lib/src/web/web_io.dart:179]: Something to do here? - If not then let's remove the comment.

**Review by matibat — APPROVED** [review]
Approving with non-blocking comments/suggestions. We can address them in a further tech_debt ticket.

**Review by matibat — APPROVED** [review]
-

**Review by matibat — APPROVED** [review]
:)

**Pull request closed without merging**